zeppelin-server/src/main/java/org/apache/zeppelin/service/InterpreterService.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zeppelin.service;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import jline.internal.Preconditions;
+import org.apache.commons.io.FileUtils;
+import org.apache.zeppelin.conf.ZeppelinConfiguration;
+import org.apache.zeppelin.dep.DependencyResolver;
+import org.apache.zeppelin.interpreter.InterpreterSettingManager;
+import org.apache.zeppelin.notebook.socket.Message;
+import org.apache.zeppelin.notebook.socket.Message.OP;
+import org.apache.zeppelin.rest.message.InterpreterInstallationRequest;
+import org.apache.zeppelin.socket.NotebookServer;
+import org.apache.zeppelin.utils.ThreadFactoryFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonatype.aether.RepositoryException;
+
+/**
+ * This class handles all of business logic for {@link org.apache.zeppelin.rest.InterpreterRestApi}
+ */
+public class InterpreterService {
```

Comment:
Do you plan to move all the functions of InterpreterRestApi to InterpreterService ?

Reply:
Yes, I think it's more general structure of web services

Reply:
Great, that is what I intended to do. See https://issues.apache.org/jira/browse/ZEPPELIN-3288
Zeppelin has 2 approach to call backend (rest api & web socket ). I think we need to refactor them to call the same api. Take this as example, all the interpreter operation should go to InterpreterService

Reply:
Since this is a service class, I don't think it make sense to put websocket stuff here. websocket should be the called of this service class. I think we can provide call back interface for this service class, then the caller can call the callback function when the service operation is done

Reply:
Agreed. Actually, I also thought it's not good. I'll think of it to make a better structure.

Reply:
I've introduced `MessageCallback` to handle it